disktail/disktail.go

```diff
@@ -0,0 +1,458 @@
+	return encoded[4 : 4+keylen], encoded[4+keylen:]
+}
+
+func (r *DiskTail[T]) storeOffset(partition int32, offset int64) error {
+	buf := make([]byte, 8)
+	binary.BigEndian.PutUint64(buf, uint64(offset))
+	return r.store.Set(offsetKeyForPartition(partition), buf, nosyncOption)
+}
+
+func (r *DiskTail[T]) getOffsets(partitions []int32) ([]int64, error) {
+	offsets := make([]int64, len(partitions))
+
+	for idx, part := range partitions {
+		val, closer, err := r.store.Get(offsetKeyForPartition(part))
+		if err != nil {
+			if err == pebble.ErrNotFound {
+				offsets[idx] = -1
+				continue
+			}
+			return nil, fmt.Errorf("error reading offset for partition %d: %w", part, err)
+		}
+		offsets[idx] = int64(binary.BigEndian.Uint64(val))
+		closer.Close()
+	}
+	return offsets, nil
+}
+
+func (r *DiskTail[T]) cleaner(ctx context.Context) error {
+
+	ticker := time.NewTicker(r.config.CleanInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+
+			// oldest timestamp based on now - max age
+			oldest := r.config.Clk.Now().Add(-r.config.MaxAge)
+			oldestKey := encodeKey(oldest, 0, 0)
+
+			// delete the whole range, starting from the minimal offsetkey
+			if err := r.store.DeleteRange(endOffsetRange, oldestKey, nosyncOption); err != nil {
+				return fmt.Errorf("error deleting range: %w", err)
+			}
+			ticker.Reset(r.config.CleanInterval)
+		}
+	}
+}
+
+type IterStats struct {
+	sync.Mutex
+	// number of items iterated
+	itemsIterated int64
+	// number of unique keys. Note that this value will be 0 if duplicates are included when
+	// iterating.
+	uniqueKeys int64
+
+	// time of the oldest item being iterated
+	oldestItem time.Time
+}
+
+func (is *IterStats) ItemsIterated() int64 {
+	is.Lock()
+	defer is.Unlock()
+
+	return is.itemsIterated
+}
+
+func (is *IterStats) UniqueKeys() int64 {
+	is.Lock()
+	defer is.Unlock()
+
+	return is.uniqueKeys
+}
+
+func (is *IterStats) OldestItem() time.Time {
+	is.Lock()
+	defer is.Unlock()
+
+	return is.oldestItem
+}
+
+type Item[T any] struct {
+	Value     T
+	Key       string
+	Timestamp time.Time
+}
+
+type HandleResult int32
+
+const (
+	Continue HandleResult = 0
+	Stop     HandleResult = 1
+)
+
+func (r *DiskTail[T]) Iterate(ctx context.Context, skipDuplicates bool, stats *IterStats, handler func(item *Item[T]) HandleResult) error {
+	if stats == nil {
+		stats = &IterStats{}
+	}
+
+	it, err := r.store.NewIter(nil)
+	if err != nil {
+		return fmt.Errorf("error creating iterator: %w", err)
+	}
+	r.store.NewBatch()
+	if !it.Last() {
+		return nil
+	}
+
+	itemHandler := handler
+
+	if skipDuplicates {
+		keys := make(map[string]struct{})
+
+		itemHandler = func(item *Item[T]) HandleResult {
+			if _, ok := keys[item.Key]; ok {
+				// skip it
+				return Continue
+			}
+			keys[item.Key] = struct{}{}
+
+			stats.Lock()
+			stats.uniqueKeys++
+			stats.Unlock()
+
+			// call original handler
+			return handler(item)
+		}
+	}
+
+	for {
```

Comment:
```suggestion
	for ctx.Err() == nil {
```
and remove the condition below

Reply:
yeah makes sense